sql/core/src/test/scala/org/apache/spark/sql/VariantShreddingSuite.scala

```diff
@@ -0,0 +1,138 @@
+    val builder = new VariantBuilder(false)
+    metadataKeys.foreach(builder.addKey)
+    builder.appendVariant(VariantBuilder.parseJson(s, false))
+    builder.result().getValue
+  }
+
+  def checkRead(path: File, expected: Seq[String]): Unit = {
+    withAllParquetReaders {
+      val df = spark.read.schema("v variant").parquet(path.getAbsolutePath)
+        .selectExpr("to_json(v)")
+      checkAnswer(df, expected.map(Row(_)))
+    }
+  }
+
+  test("scalar types rebuild") {
+    val scalarTypes = Array(
+      BooleanType, ByteType, ShortType, IntegerType, LongType, FloatType, DoubleType,
+      TimestampType, TimestampNTZType, DateType,
+      StringType, BinaryType,
+      DecimalType(9, 3), DecimalType(18, 6), DecimalType(22, 9))
+    val obj = StructType(scalarTypes.zipWithIndex.map { case (t, i) =>
+      StructField(i.toString, StructType(Array(StructField("typed_value", t))))
+    })
+    val v = StructType(Array(
+      StructField("typed_value", obj),
+      StructField("value", BinaryType),
+      StructField("metadata", BinaryType)
+    ))
+
+    val values = Seq[Any](
+      true, 1.toByte, 2.toShort, 3, 4L, 5.5F, 6.6,
+      new Timestamp(7), LocalDateTime.of(1, 1, 1, 0, 0, 8, 0), new Date(9),
+      "str10", Array[Byte](11),
+      Decimal("12.12"), Decimal("13.13"), Decimal("14.14"))
+    val row = Row(Row.fromSeq(values.map(Row(_))), null,
+      metadata(scalarTypes.indices.map(_.toString)))
+
+    withTempPath { path =>
+      spark.createDataFrame(spark.sparkContext.parallelize(Seq(Row(row))),
+          StructType(Array(StructField("v", v))))
+        .write.parquet(path.getAbsolutePath)
+      for (tz <- Seq("Etc/UTC", "America/Los_Angeles")) {
+        withSQLConf(SQLConf.SESSION_LOCAL_TIMEZONE.key -> tz) {
+          val timestamp = if (tz == "Etc/UTC") {
+            "1970-01-01 00:00:00.007+00:00"
+          } else {
+            "1969-12-31 16:00:00.007-08:00"
+          }
+          checkRead(path, Seq(
+            """{"0":true,"1":1,"10":"str10","11":"Cw==","12":12.12,"13":13.13,"14":14.14,""" +
+              s""""2":2,"3":3,"4":4,"5":5.5,"6":6.6,"7":"$timestamp",""" +
+              """"8":"0001-01-01 00:00:08","9":"1969-12-31"}"""))
+        }
+      }
+    }
+  }
+
+  test("object rebuild") {
+    val schema = StructType.fromDDL("v struct<metadata binary, value binary, " +
+      "typed_value struct<b struct<typed_value int, value binary>, " +
+      "d struct<typed_value int, value binary>>>")
+    withTempPath { path =>
+      spark.createDataFrame(spark.sparkContext.parallelize(Seq(
+          Row(metadata(Seq("b", "d")), null, Row(Row(1, null), Row(null, null))),
+          Row(metadata(Seq("b", "d")), null, Row(Row(1, null), Row(null, value("null")))),
+          Row(metadata(Seq("a", "b", "c", "d")),
+            shreddedValue("""{"a": 1, "c": 3}""", Seq("a", "b", "c", "d")),
+            Row(Row(2, null), Row(null, value("4")))),
```

Comment:
Just to make sure I understand, would the result here be any different if the value "4" was put into the typed_value field of `d`? Is this an example where shredding made a suboptimal but valid decision?

Reply:
There is no difference. Putting 4 in variant value is also valid input data, though the writer may not write such data.